Rewrite of Connection Migration logic

janaiyengar opened the pull request:
This is my first complete pass at a rewrite as we discussed earlier today, PTAL.

Commit:
initial changes

Commit:
initial overview and some nits

Commit:
wrapped up path probing section

Commit:
section on committing

Commit:
committing before I start pruning

Commit:
first complete pass

Commit:
change abandon behavior

tfpauly commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1461,126 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+frame MUST contain at least 12 randomly generated octets {{?RFC4086}}, making
+the payload sufficiently hard to guess.
+
+An endpoint SHOULD not generate two PATH_CHALLENGE frames with the same payload.
```
Does this mean that we shouldn't have the same PATH_CHALLENGE for two different paths, or that if we send multiple challenges on the same path, they also need to be different?

tfpauly replied:
And this should be SHOULD NOT, rather than SHOULD not.

tfpauly replied:
Based on below text, it seems like it is the latter case.

janaiyengar replied:
I figured unique for all paths, since having the same random bytes used for two different paths can also get confusing, and because it's fairly trivial to not have them be the same. Do you think it needs clarifying? (I'll fix the SHOULD NOT)

tfpauly replied:
Nope, it's fine as is I think.

janaiyengar replied:
I've changed some text here.

tfpauly replied:
Thanks, this looks good!

tfpauly commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1461,126 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+
+### Client Behavior {{migration-client}}
+
+When a new local address is available, a client MUST first send a packet
```
Saying "When a new local address is available" and then describing a MUST about client behavior may be a bit misleading, indicating something about the client needing to migrate when a new address is available (I know that's not what it says, though).

Can we instead say "When the client intends to send packets from a new local address" or "When the client intends to use a new local address"

janaiyengar replied:
Done.

tfpauly commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1461,126 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+period, but not receiving a PATH_RESPONSE may be indicative of a PMTU detection
+failure.
+
+A client may receive a PATH challenge from a server that is validating the
```
PATH challenge -> PATH_CHALLENGE

janaiyengar replied:
Done

tfpauly commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1461,126 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+
+A client may receive a PATH challenge from a server that is validating the
+client's ownership of an address. A client responds to this frame by echoing the
+data from the PATH_CHALLENGE frame in a PATH_RESPONSE frame.  This response MAY
```
To be super clear, maybe say "datas from the server's PATH_CHALLENGE frame"

janaiyengar replied:
Done

tfpauly commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1461,126 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+When a server receives a PATH_CHALLENGE frame from a client, it MUST respond by
+copying the included data in a PATH_RESPONSE frame, and send it in a packet
+padded to the size of the packet carrying the PATH_CHALLENGE. This PATH_RESPONSE
+MUST be sent to the address that the PATH_CHALLENGE was received from.
```
Minor wordsmithing:

When a server receives a PATH_CHALLENGE frame from a client, it MUST generate a PATH_RESPONSE frame that contains the same data as the challenge, and send it in a packet padded to the size of the client's packet that included the PATH_CHALLENGE. This PATH_RESPONSE MUST be sent to the address from which the PATH_CHALLENGE was received.

janaiyengar replied:
Done

tfpauly commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1461,126 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+was sent as valid.
+
+A server MAY send additional PATH_CHALLENGE frames to handle packet loss,
+subject to the congestion controller's limits. The server MUST use fresh random
```
Why is this a MUST for server, and a SHOULD for clients?

janaiyengar replied:
Because there are potential security implications for the server but not for the client.

tfpauly replied:
Should we have a line explaining what the security concern is?

janaiyengar replied:
I've added some text

tfpauly commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1461,126 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+PATH_RESPONSE, or PADDING from a client address with a packet number that is the
+largest seen thus far on the connection, the server commits to this client
+address. The server MUST send all subsequent packets, with the exception of
+those containing PATH_RESPONSE and PATH_CHALLENGE frames, to this address.
```
Do we want to explicitly mention the converse of this statement (to make sure implementations do it right), that just sending PATH_CHALLENGE/PATH_RESPONSE/PADDING only probes a path without changing where the server sends data back to the client?

janaiyengar replied:
Done.

tfpauly replied:
Looks perfect

Commit:
more structure and some more text

janaiyengar commented:
Ok, I've addressed those comments and I've added some more structure and text, PTAL. Ideally I'd like to send out something to folks in our small group by the end of the weekend or early Monday, and send out the PR to the QUIC wg by Tuesday. SG?

tfpauly commented:
Text looks good to me!

We still need to adjust the frame types to bring PING back, right? Eric, were you going to do that? I can work on that this weekend if that's not covered already

Commit:
editorial changes and added section on loss detection

Commit:
added PING frame back in

janaiyengar commented:
Ok, I've added a section on loss detection and congestion control, and I made some editorial nits to the text.  I also added the PING frame back in. PTAL when you get a chance. I'd like to ship this out tomorrow to our small group if you're both ok with this.

erickinnear commented:
Thanks for grabbing the PING frame back in, hadn't gotten to that yet. Taking a look at the rest of it now!

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
-also find that a path suddenly becomes inoperable, such as when a user
-unplugs a cable or other network state changes, and may elect to migrate
-without probing since it has no alternative if it wishes to maintain the
+QUIC connections are identified by 64-bit Connection IDs. QUIC's connection IDs
```
`Connection IDs` the first time, but then `connection IDs` the second. Looks like we use lower case elsewhere?

janaiyengar replied:
fixed

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+owns an address, or to verify the PMTU to/from an address.
+
+A server validates a client's address by sending a PATH_CHALLENGE frame
+containing a payload that is hard to guess to that address.  The server
```
In path validation we previously defined what was considered "hard to guess". Do we want to pull that in here?

erickinnear replied:
Ah, I see you did that below where it's relevant for the server, that's a good spot.

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+address to the server by sending a PMTU-sized packet carrying a PATH_CHALLENGE
+frame. The client considers the PMTU verified when a PATH_RESPONSE frame
+containing the same payload sent by the server to the new client address is
+received.
```
Note that this PATH_RESPONSE frame is also PMTU-sized? 
We require that the server pad, but if the client receives a non-padded packet then that would just be a connection error?

janaiyengar replied:
Well, the client doesn't know the MTU on the reverse path, so it can't detect whether it's PMTU sized or not... besides, this is not a security concern, so the client doesn't need to do anything special, IMO. I've added some more text below in server behavior.

erickinnear replied:
Ah, fair enough.

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+A client triggers connection migration to a new address by sending a
+PATH_CHALLENGE as a probe or by switching to sending all packets to the new
+address. The client SHOULD do PMTU verification before considering migration
+complete. The client also participates in a validation that the server may
```
This came out nicely!

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+A client triggers connection migration to a new address by sending a
+PATH_CHALLENGE as a probe or by switching to sending all packets to the new
+address. The client SHOULD do PMTU verification before considering migration
+complete. The client also participates in a validation that the server may
```
Nit: `in a validation` -> `in address validation`

janaiyengar replied:
done

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+validation token to the client ({{address-validation}}).
+
+
+#### Committing to a New Address
```
Organization-ally, we have: 
```
Client: 
1. What it should do
2. How it responds to what the server does
Server: 
1. How it responds to what the client does
2. What it should do
Okay we're done now, let's switch!
```
This inversion works very nicely in terms of matching thinking from each side of the connection at the same time as handling two separate operations from each end.

janaiyengar replied:
Yeah, that was my thinking -- glad it seems to work well.

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+
+A client MAY send additional PATH_CHALLENGE frames to handle packet loss or to
+get more measurements on the new network path, subject to the congestion
+controller's limits.
```
Link to `### Loss Detection and Congestion Control`?

janaiyengar replied:
I think the recovery doc is more appropriate; added that

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+
+When the client receives a PATH_RESPONSE from the server and verifies that the
+included data matches that sent by the client in a previously sent
+PATH_CHALLENGE, the server is considered reachable from the new address and the
```
Minor: `that sent by the client in a previously sent` reads as awkward to me. 
How about: 
`When the client receives a PATH_RESPONSE frame from the server and verifies that the included data matches the data it sent in a previous PATH_CHALLENGE, the...`

janaiyengar replied:
done

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+unvalidated address.
+
+
+#### Responding to Client's Probe
```
Nit: `to` -> `to the`?

janaiyengar replied:
changed to "to a"

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+failure.
+
+
+#### Responding to Server's Validation
```
Nit: `to` -> `to the`?

janaiyengar replied:
changed to "to a"

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+
+The server responds to any probes sent by the client from a new address,
+participates in the client's PMTU verification, and MUST validate that the
+client owns its new address.  The server MAY validate any client's address at
```
Minor: `its` might be unclear here? Perhaps `MUST validate that the client owns the new address seen by the server` or something along those lines? Not a big deal, would be fine to leave alone.

janaiyengar replied:
changed to "that the client owns the new address", which works here I think

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+valid, the server MUST send a PATH_CHALLENGE frame with its own validation data
+to this address to verify the client's ownership of it. The server's
+PATH_CHALLENGE MAY be bundled with other frames in a packet that does not need
+to be padded.
```
This reads funny to me. We're trying to say "if you want, you can send this in any other packet, even if that packet does need to be padded, as long as you still pad that packet", right? 
i.e. 
`The server's PATH_CHALLENGE MAY be bundled with other frames in a packet, since it does not place any requirements on the size of its containing packet.`

erickinnear replied:
Or even `since it does not require additional padding` as we said above in the client section.

janaiyengar replied:
done -- used second suggestion

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+PATH_RESPONSE, or PADDING from a client address with a packet number that is the
+largest seen thus far on the connection, the server "commits" to this client
+address. The server MUST send all subsequent packets, with the exception of
+those containing PATH_RESPONSE and PATH_CHALLENGE frames, to this address.
```
The intent here is to require all frames except those (and padding) frames to go to the "main" address for the client, right? 
In that case, should we say `containing only PATH_RESPONSE, PATH_CHALLENGE, or PADDING frames`? Otherwise we're letting the difference between frames and packets prevent this from actually requiring that.

janaiyengar replied:
Ok, I've rephrased this part a bit. Take a look.

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+Upon committing to a new address, the server SHOULD reset its congestion
+controller and roundtrip time estimator to start new estimations.
+
+
```
Irrelevant: Extra line here on purpose?

janaiyengar replied:
fixed

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -2236,40 +2213,29 @@ than it has sent, unless this is a result of a change in the initial limits (see
-migration.
+Endpoints can use PING frames (type=0x07) to verify that their peers are still
+alive or to check reachability to the peer.  The PING frame contains no
+additional fields.  The receiver of a PING frame simply needs to acknowledge the
```
`The PING frame contains no additional fields.`
Worth having this twice if we're just going to say it again right below?

janaiyengar replied:
Hm, that must've crept in -  fixed.

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -2394,18 +2360,56 @@ Application Error Code:
 
+## PATH_CHALLENGE Frame {#frame-path-challenge}
+
+Endpoints can use PATH_CHALLENGE frames (type=0x07) to check reachability to the
```
PATH_CHALLENGE is now type `0x0e`

janaiyengar replied:
fixed

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -2394,18 +2360,56 @@ Application Error Code:
+The recipient of this frame MUST generate a PATH_RESPONSE frame
+({{frame-path-response}}) containing the same Data.  An endpoint that receives a
+PATH_CHALLENGE frame containing an empty payload MUST generate a connection
+error of type FRAME_ERROR, indicating the PATH_CHALLENGE frame (that is, 0x0d).
```
`0x0e`?

janaiyengar replied:
sigh, fixed.

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -2394,18 +2360,56 @@ Application Error Code:
+
 ## PATH_RESPONSE Frame {#frame-path-response}
 
 The PATH_RESPONSE frame (type=0x0d) is sent in response to a PATH_CHALLENGE
```
`0x0f`

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -2394,18 +2360,56 @@ Application Error Code:
-generate a connection error of type UNSOLICITED_PATH_RESPONSE.
+An endpoint that receives a PATH_RESPONSE frame containing an empty payload MUST
+generate a connection error of type FRAME_ERROR, indicating the PATH_RESPONSE
+frame (that is, 0x0e).  If the content of a PATH_RESPONSE frame does not match
```
`0x0f`

janaiyengar replied:
wow, I was really on something here

erickinnear replied:
:)

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -3793,8 +3796,8 @@ Issue and pull request numbers are listed with a leading octothorp.
-- Replace PING and PONG frames with PATH_CHALLENGE and PATH_RESPONSE frames,
-  clarify connection migration text (#000)
+- Added PATH_CHALLENGE and PATH_RESPONSE frames, removed PING with Data,
+  and rewrote connection migration (#000)
```
Probably worth mentioning that we also removed the PONG frame

janaiyengar replied:
good point, done.

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+quiescence.
+
+
+### Server Behavior {{migration-server}}
```
Shouldn't this be `{#migration-server}`?

janaiyengar replied:
sigh, yes. fixed.

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+received.
+
+
+### Client Behavior {{migration-client}}
```
Shouldn't this be {#migration-client}?

janaiyengar replied:
yup, fixed

erickinnear commented on draft-ietf-quic-transport.md:
```diff
@@ -1461,127 +1462,215 @@ connection if the integrity check fails with a PROTOCOL_VIOLATION error code.
+received for a prior one.
+
+
+#### Validating an Unvalidated Client Address
```
Let's call this `{#migration-validate}` since we refer to that elsewhere in the document?

janaiyengar replied:
done

erickinnear reviewed (commented):
Looks great so far! Thanks for putting this together and cleaning it up!

Commit:
ekinnear's comments and some wordsmithing
